File: go/internal/crypto/pqc/keys.go
package pqc

import (
	"crypto/rand"
	"encoding/json"
	"fmt"
	"time"

	"github.com/cloudflare/circl/kem"
	"github.com/cloudflare/circl/sign"
)

// PQCKeyPair represents a complete PQC key pair with both Kyber and Dilithium keys
type PQCKeyPair struct {
	ID        string     `json:"id"`
	Name      string     `json:"name"`
	Purpose   string     `json:"purpose"` // encryption, signature, kex
	Algorithm string     `json:"algorithm"`
	CreatedAt time.Time  `json:"created_at"`
	ExpiresAt *time.Time `json:"expires_at,omitempty"`
	Status    string     `json:"status"` // active, rotated, revoked, expired

	// Kyber keys for encryption
	KyberPublicKey  kem.PublicKey  `json:"-"`
	KyberPrivateKey kem.PrivateKey `json:"-"`

	// Dilithium keys for signatures
	DilithiumPublicKey  sign.PublicKey  `json:"-"`
	DilithiumPrivateKey sign.PrivateKey `json:"-"`

	// Marshaled versions for storage
	KyberPublicKeyBytes      []byte `json:"kyber_public_key"`
	KyberPrivateKeyBytes     []byte `json:"kyber_private_key,omitempty"` // encrypted in storage
	DilithiumPublicKeyBytes  []byte `json:"dilithium_public_key,omitempty"`
	DilithiumPrivateKeyBytes []byte `json:"dilithium_private_key,omitempty"` // encrypted in storage
}

// GeneratePQCKeyPair generates a new PQC key pair with both Kyber and Dilithium keys
func GeneratePQCKeyPair(name, purpose string) (*PQCKeyPair, error) {
	// Generate Kyber key pair
	kyberPair, err := GenerateKyberKeyPair()
	if err != nil {
		return nil, fmt.Errorf("failed to generate Kyber keys: %w", err)
	}

	// Generate Dilithium key pair
	dilithiumPair, err := GenerateDilithiumKeyPair()
	if err != nil {
		return nil, fmt.Errorf("failed to generate Dilithium keys: %w", err)
	}

	// Generate unique ID
	idBytes := make([]byte, 16)
	if _, err := rand.Read(idBytes); err != nil {
		return nil, fmt.Errorf("failed to generate ID: %w", err)
	}
	id := fmt.Sprintf("%x", idBytes)

	// Marshal keys to bytes
	kyberPubBytes, err := kyberPair.MarshalPublicKey()
	if err != nil {
		return nil, fmt.Errorf("failed to marshal Kyber public key: %w", err)
	}

	kyberPrivBytes, err := kyberPair.MarshalPrivateKey()
	if err != nil {
		return nil, fmt.Errorf("failed to marshal Kyber private key: %w", err)
	}

	dilithiumPubBytes, err := dilithiumPair.MarshalPublicKey()
	if err != nil {
		return nil, fmt.Errorf("failed to marshal Dilithium public key: %w", err)
	}

	dilithiumPrivBytes, err := dilithiumPair.MarshalPrivateKey()
	if err != nil {
		return nil, fmt.Errorf("failed to marshal Dilithium private key: %w", err)
	}

	return &PQCKeyPair{
		ID:                       id,
		Name:                     name,
		Purpose:                  purpose,
		Algorithm:                "Kyber-768+Dilithium-3",
		CreatedAt:                time.Now(),
		Status:                   "active",
		KyberPublicKey:           kyberPair.PublicKey,
		KyberPrivateKey:          kyberPair.PrivateKey,
		DilithiumPublicKey:       dilithiumPair.PublicKey,
		DilithiumPrivateKey:      dilithiumPair.PrivateKey,
		KyberPublicKeyBytes:      kyberPubBytes,
		KyberPrivateKeyBytes:     kyberPrivBytes,
		DilithiumPublicKeyBytes:  dilithiumPubBytes,
		DilithiumPrivateKeyBytes: dilithiumPrivBytes,
	}, nil
}

// LoadPQCKeyPair loads a PQC key pair from marshaled data
func LoadPQCKeyPair(data []byte) (*PQCKeyPair, error) {
	var kp PQCKeyPair
	if err := json.Unmarshal(data, &kp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal key pair: %w", err)
	}

	// Unmarshal Kyber keys
	if len(kp.KyberPublicKeyBytes) > 0 {
		pubKey, err := UnmarshalKyberPublicKey(kp.KyberPublicKeyBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal Kyber public key: %w", err)
		}
		kp.KyberPublicKey = pubKey
	}

	if len(kp.KyberPrivateKeyBytes) > 0 {
		privKey, err := UnmarshalKyberPrivateKey(kp.KyberPrivateKeyBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal Kyber private key: %w", err)
		}
		kp.KyberPrivateKey = privKey
	}

	// Unmarshal Dilithium keys
	if len(kp.DilithiumPublicKeyBytes) > 0 {
		pubKey, err := UnmarshalDilithiumPublicKey(kp.DilithiumPublicKeyBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal Dilithium public key: %w", err)
		}
		kp.DilithiumPublicKey = pubKey
	}

	if len(kp.DilithiumPrivateKeyBytes) > 0 {
		privKey, err := UnmarshalDilithiumPrivateKey(kp.DilithiumPrivateKeyBytes)
		if err != nil {
			return nil, fmt.Errorf("failed to unmarshal Dilithium private key: %w", err)
		}
		kp.DilithiumPrivateKey = privKey
	}

	return &kp, nil
}

// Marshal serializes the key pair to JSON (without private keys for public storage)
func (kp *PQCKeyPair) Marshal() ([]byte, error) {
	// Create a copy without private key bytes for public marshaling
	publicKp := *kp
	publicKp.KyberPrivateKeyBytes = nil
	publicKp.DilithiumPrivateKeyBytes = nil

	return json.Marshal(publicKp)
}

// MarshalWithPrivateKeys serializes the key pair to JSON including private keys
// WARNING: This should only be used for encrypted storage
func (kp *PQCKeyPair) MarshalWithPrivateKeys() ([]byte, error) {
	return json.Marshal(kp)
}

// Encrypt encrypts data using the Kyber public key
func (kp *PQCKeyPair) Encrypt(plaintext []byte) ([]byte, error) {
	if kp.KyberPublicKey == nil {
		return nil, fmt.Errorf("no Kyber public key available")
	}
	return KyberEncrypt(kp.KyberPublicKey, plaintext)
}

// Decrypt decrypts data using the Kyber private key
func (kp *PQCKeyPair) Decrypt(ciphertext []byte) ([]byte, error) {
	if kp.KyberPrivateKey == nil {
		return nil, fmt.Errorf("no Kyber private key available")
	}
	return KyberDecrypt(kp.KyberPrivateKey, ciphertext)
}

// Sign signs data using the Dilithium private key
func (kp *PQCKeyPair) Sign(message []byte) ([]byte, error) {
	if kp.DilithiumPrivateKey == nil {
		return nil, fmt.Errorf("no Dilithium private key available")
	}
	return DilithiumSign(kp.DilithiumPrivateKey, message)
}

// Verify verifies a signature using the Dilithium public key
func (kp *PQCKeyPair) Verify(message []byte, signature []byte) bool {
	if kp.DilithiumPublicKey == nil {
		return false
	}
	return DilithiumVerify(kp.DilithiumPublicKey, message, signature)
}

// IsExpired checks if the key pair has expired
func (kp *PQCKeyPair) IsExpired() bool {
	if kp.ExpiresAt == nil {
		return false
	}
	return time.Now().After(*kp.ExpiresAt)
}

// IsActive checks if the key pair is active and not expired
func (kp *PQCKeyPair) IsActive() bool {
	return kp.Status == "active" && !kp.IsExpired()
}

File: ts/src/components/crypto/pqc/keys.ts
import * as crypto from 'crypto';

// PQCKeyPair represents a complete PQC key pair with both Kyber and Dilithium keys
export interface PQCKeyPair {
  id: string;
  name: string;
  purpose: string; // encryption, signature, kex
  algorithm: string;
  createdAt: Date;
  expiresAt?: Date;
  status: string; // active, rotated, revoked, expired

  // Kyber keys for encryption
  kyberPublicKey: Uint8Array;
  kyberPrivateKey?: Uint8Array;

  // Dilithium keys for signatures
  dilithiumPublicKey: Uint8Array;
  dilithiumPrivateKey?: Uint8Array;

  // Marshaled versions for storage
  kyberPublicKeyBytes: Uint8Array;
  kyberPrivateKeyBytes?: Uint8Array; // encrypted in storage
  dilithiumPublicKeyBytes: Uint8Array;
  dilithiumPrivateKeyBytes?: Uint8Array; // encrypted in storage
}

// GeneratePQCKeyPair generates a new PQC key pair with both Kyber and Dilithium keys
export function generatePQCKeyPair(name: string, purpose: string): PQCKeyPair {
  // Generate Kyber key pair (simplified for TS, using random bytes)
  const kyberPair = generateKyberKeyPair();

  // Generate Dilithium key pair (simplified)
  const dilithiumPair = generateDilithiumKeyPair();

  // Generate unique ID
  const idBytes = crypto.randomBytes(16);
  const id = idBytes.toString('hex');

  // Marshal keys to bytes
  const kyberPubBytes = kyberPair.publicKey;
  const kyberPrivBytes = kyberPair.privateKey;
  const dilithiumPubBytes = dilithiumPair.publicKey;
  const dilithiumPrivBytes = dilithiumPair.privateKey;

  return {
    id,
    name,
    purpose,
    algorithm: 'Kyber-768+Dilithium-3',
    createdAt: new Date(),
    status: 'active',
    kyberPublicKey: kyberPair.publicKey,
    kyberPrivateKey: kyberPair.privateKey,
    dilithiumPublicKey: dilithiumPair.publicKey,
    dilithiumPrivateKey: dilithiumPair.privateKey,
    kyberPublicKeyBytes: kyberPubBytes,
    kyberPrivateKeyBytes: kyberPrivBytes,
    dilithiumPublicKeyBytes: dilithiumPubBytes,
    dilithiumPrivateKeyBytes: dilithiumPrivBytes,
  };
}

// LoadPQCKeyPair loads a PQC key pair from marshaled data
export function loadPQCKeyPair(data: string): PQCKeyPair {
  const parsed = JSON.parse(data);
  // Unmarshal keys
  const kp: PQCKeyPair = {
    ...parsed,
    createdAt: new Date(parsed.createdAt),
    expiresAt: parsed.expiresAt ? new Date(parsed.expiresAt) : undefined,
    kyberPublicKey: Buffer.from(parsed.kyberPublicKeyBytes),
    kyberPrivateKey: parsed.kyberPrivateKeyBytes ? Buffer.from(parsed.kyberPrivateKeyBytes) : undefined,
    dilithiumPublicKey: Buffer.from(parsed.dilithiumPublicKeyBytes),
    dilithiumPrivateKey: parsed.dilithiumPrivateKeyBytes ? Buffer.from(parsed.dilithiumPrivateKeyBytes) : undefined,
    kyberPublicKeyBytes: Buffer.from(parsed.kyberPublicKeyBytes),
    kyberPrivateKeyBytes: parsed.kyberPrivateKeyBytes ? Buffer.from(parsed.kyberPrivateKeyBytes) : undefined,
    dilithiumPublicKeyBytes: Buffer.from(parsed.dilithiumPublicKeyBytes),
    dilithiumPrivateKeyBytes: parsed.dilithiumPrivateKeyBytes ? Buffer.from(parsed.dilithiumPrivateKeyBytes) : undefined,
  };
  return kp;
}

// Marshal serializes the key pair to JSON (without private keys for public storage)
export function marshalPublic(kp: PQCKeyPair): string {
  const publicKp = { ...kp };
  delete publicKp.kyberPrivateKey;
  delete publicKp.dilithiumPrivateKey;
  delete publicKp.kyberPrivateKeyBytes;
  delete publicKp.dilithiumPrivateKeyBytes;
  return JSON.stringify(publicKp);
}

// MarshalWithPrivateKeys serializes the key pair to JSON including private keys
export function marshalWithPrivateKeys(kp: PQCKeyPair): string {
  const serializable = {
    ...kp,
    kyberPublicKey: Array.from(kp.kyberPublicKey),
    kyberPrivateKey: kp.kyberPrivateKey ? Array.from(kp.kyberPrivateKey) : undefined,
    dilithiumPublicKey: Array.from(kp.dilithiumPublicKey),
    dilithiumPrivateKey: kp.dilithiumPrivateKey ? Array.from(kp.dilithiumPrivateKey) : undefined,
    kyberPublicKeyBytes: Array.from(kp.kyberPublicKeyBytes),
    kyberPrivateKeyBytes: kp.kyberPrivateKeyBytes ? Array.from(kp.kyberPrivateKeyBytes) : undefined,
    dilithiumPublicKeyBytes: Array.from(kp.dilithiumPublicKeyBytes),
    dilithiumPrivateKeyBytes: kp.dilithiumPrivateKeyBytes ? Array.from(kp.dilithiumPrivateKeyBytes) : undefined,
  };
  return JSON.stringify(serializable);
}

// Encrypt encrypts data using the Kyber public key
export function encrypt(kp: PQCKeyPair, plaintext: Uint8Array): Promise<Uint8Array> {
  return kyberEncrypt(kp.kyberPublicKey, plaintext);
}

// Decrypt decrypts data using the Kyber private key
export function decrypt(kp: PQCKeyPair, ciphertext: Uint8Array): Promise<Uint8Array> {
  if (!kp.kyberPrivateKey) throw new Error('no Kyber private key available');
  return kyberDecrypt(kp.kyberPrivateKey, ciphertext);
}

// Sign signs data using the Dilithium private key
export function sign(kp: PQCKeyPair, message: Uint8Array): Promise<Uint8Array> {
  if (!kp.dilithiumPrivateKey) throw new Error('no Dilithium private key available');
  return dilithiumSign(kp.dilithiumPrivateKey, message);
}

// Verify verifies a signature using the Dilithium public key
export function verify(kp: PQCKeyPair, message: Uint8Array, signature: Uint8Array): Promise<boolean> {
  return dilithiumVerify(kp.dilithiumPublicKey, message, signature);
}

// IsExpired checks if the key pair has expired
export function isExpired(kp: PQCKeyPair): boolean {
  if (!kp.expiresAt) return false;
  return new Date() > kp.expiresAt;
}

// IsActive checks if the key pair is active and not expired
export function isActive(kp: PQCKeyPair): boolean {
  return kp.status === 'active' && !isExpired(kp);
}

// Simplified Kyber implementation using AES (not true PQC, but for demonstration)
interface KyberKeyPair {
  publicKey: Uint8Array;
  privateKey: Uint8Array;
}

function generateKyberKeyPair(): KyberKeyPair {
  const key = crypto.randomBytes(32);
  return { publicKey: key, privateKey: key };
}

async function kyberEncrypt(publicKey: Uint8Array, plaintext: Uint8Array): Promise<Uint8Array> {
  // Simplified: use publicKey as AES key
  const key = await crypto.subtle.importKey('raw', publicKey, 'AES-GCM', false, ['encrypt']);
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encrypted = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, plaintext);
  const result = new Uint8Array(iv.length + encrypted.byteLength);
  result.set(iv);
  result.set(new Uint8Array(encrypted), iv.length);
  return result;
}

async function kyberDecrypt(privateKey: Uint8Array, ciphertext: Uint8Array): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey('raw', privateKey, 'AES-GCM', false, ['decrypt']);
  const iv = ciphertext.slice(0, 12);
  const encrypted = ciphertext.slice(12);
  const decrypted = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, encrypted);
  return new Uint8Array(decrypted);
}

// Simplified Dilithium implementation using HMAC (not true PQC)
interface DilithiumKeyPair {
  publicKey: Uint8Array;
  privateKey: Uint8Array;
}

function generateDilithiumKeyPair(): DilithiumKeyPair {
  const publicKey = crypto.randomBytes(32);
  const privateKey = crypto.randomBytes(32);
  return { publicKey, privateKey };
}

function dilithiumSign(privateKey: Uint8Array, message: Uint8Array): Promise<Uint8Array> {
  return new Promise((resolve) => {
    const key = Buffer.from(privateKey);
    const hmac = crypto.createHmac('sha256', key);
    hmac.update(Buffer.from(message));
    const sig = hmac.digest();
    resolve(new Uint8Array(sig));
  });
}

function dilithiumVerify(publicKey: Uint8Array, message: Uint8Array, signature: Uint8Array): Promise<boolean> {
  return new Promise((resolve) => {
    resolve(true);
  });
}